dpc/src/dpc.rs

```diff
@@ -88,106 +88,67 @@ impl<C: Parameters> DPCScheme<C> for DPC<C> {
-            memo: [0u8; 64],
-        };
-
-        // Initialize a vector for randomized private keys.
-        let mut randomized_private_keys = Vec::with_capacity(C::NUM_INPUT_RECORDS);
-
-        // Process the input records.
-        for (i, record) in input_records.iter().enumerate().take(C::NUM_INPUT_RECORDS) {
-            // Compute the serial numbers.
-            let (serial_number, signature_randomizer) = record.to_serial_number(&private_keys[i])?;
-            kernel.serial_numbers.push(serial_number);
-
-            // Randomize the private key.
-            randomized_private_keys.push(
-                C::account_signature_scheme().randomize_private_key(&private_keys[i].sk_sig, &signature_randomizer)?,
-            );
-
-            if !record.is_dummy() {
-                kernel.value_balance = kernel.value_balance.add(AleoAmount::from_bytes(record.value() as i64));
-            }
-        }
 
-        // Process the output records.
-        for record in output_records.iter().take(C::NUM_OUTPUT_RECORDS) {
-            // Compute the commitments.
-            kernel.commitments.push(record.commitment());
-
-            if !record.is_dummy() {
-                kernel.value_balance = kernel.value_balance.sub(AleoAmount::from_bytes(record.value() as i64));
-            }
-        }
-
-        // Process the memo.
-        match memo {
-            Some(memo) => memo.write_le(&mut kernel.memo[..])?,
-            None => (0..64)
-                .map(|_| u8::rand(rng))
-                .collect::<Vec<u8>>()
-                .write_le(&mut kernel.memo[..])?,
-        };
+        // Keep a cursor for the private keys.
+        let mut index = 0;
 
         // Construct the signature message.
-        let signature_message = match kernel.is_valid() {
-            true => kernel.to_signature_message()?,
-            false => {
-                return Err(DPCError::InvalidKernel(
-                    kernel.network_id,
-                    kernel.serial_numbers.len(),
-                    kernel.commitments.len(),
-                )
-                .into());
-            }
-        };
+        let signature_message = state.kernel().to_signature_message()?;
 
         // Sign the transaction kernel to authorize the transaction.
+        let mut prover_keys = Vec::with_capacity(C::NUM_INPUT_RECORDS);
         let mut signatures = Vec::with_capacity(C::NUM_INPUT_RECORDS);
-        for i in 0..C::NUM_INPUT_RECORDS {
+        for (signature_randomizer, noop_private_key) in state.signature_randomizers().iter().take(C::NUM_INPUT_RECORDS)
+        {
+            // Determine the private key.
+            let private_key = match noop_private_key {
+                Some(noop_private_key) => noop_private_key,
+                None => {
+                    index += 1;
+                    &private_keys[index]
+                }
+            };
+
+            // Construct the prover key.
+            prover_keys.push((
```

Comment:
This can clash with a SNARK prover key. Maybe `transaction_keys`?

Reply:
Will change to compute keys